rascal-lsp/src/main/rascal/util/refactor/Rename.rsc

```diff
@@ -0,0 +1,193 @@
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+@bootstrapParser
+module util::refactor::Rename
+
+import ParseTree;
+import Set;
+import util::Maybe;
+import util::Monitor;
+import util::Reflective;
+
+import IO;
+
+import util::refactor::Exception;
+import util::refactor::TextEdits;
+extend util::refactor::WorkspaceInfo;
+
+import analysis::typepal::TModel;
+
+@synopsis{ To be extended by implementations. }
+data Cursor;
+
+@synopsis{ Return type for rename refactoring. }
+alias Edits = tuple[list[DocumentEdit], map[ChangeAnnotationId, ChangeAnnotation]];
+@synopsis{ Return type for rename validity checks. }
+alias CheckResult = set[IllegalRenameReason];
+@synopsis {Get a path config for a project folder. }
+alias PathConfigF = PathConfig(loc);
+@synopsis { A rename validity check that does not require type check information. }
+alias PreCheckF = CheckResult(Tree cursorT, str newName, set[loc] workspaceFolders, PathConfigF getPathConfig);
+@synopsis{ A function that checks whether the produced rename candidates are valid. }
+alias PostCheckF = CheckResult(TModel tm, loc moduleLoc, str newName, set[RenameLocation] defs, set[RenameLocation] uses);
+@synopsis{ A function that selects project files to load. }
+alias PreFilesF = ProjectFiles(Tree cursorT, set[loc] workspaceFolders, PathConfigF getPathConfig);
+alias AllFilesF = ProjectFiles(TModel tm, Cursor cur, set[loc] workspaceFolders, PathConfigF getPathConfig);
+@synopsis{ A function that, given a set of file locs and a path config, returns a set of TModels. }
+alias TModelsF = set[TModel](set[loc] fs, PathConfig pcfg);
+@synopsis{ A function that, given a TModel and the Tree under the cursor, returns a Cursor. }
+alias CursorF = Cursor(TModel tm, Tree cursorT);
+@synopsis{ A function that determines rename candidates. }
+alias DefsUsesF = DefsUsesRenames(TModel tm, Cursor cur, ChangeAnnotationRegister registerChangeAnnotation, PathConfigF getPathConfig);
+@synopsis{ A function that finds the location of names in candidates. }
+alias FindNamesF = rel[loc, loc](loc l, set[loc] candidates);
+@synopsis{ Escape a name. }
+alias EscapeNameF = str(str name);
+@synopsis{ The type of the rename function, as expected by the language server. }
+alias RenameSymbolF = Edits(Tree cursorT, str newName, set[loc] workspaceFolders, PathConfigF getPathConfig);
+
+// TODO Make parameter
+str MANDATORY_CHANGE_DESCRIPTION = "These changes are required for a correct renaming. They can be previewed here, but it is not advised to disable them.";
+
+RenameSymbolF renameSymbolFramework(
+    PreCheckF preCheck
+  , PostCheckF postCheck
+  , PreFilesF preFiles
+  , AllFilesF allFiles
+  , TModelsF getTModels
+  , EscapeNameF escapeName
+  , CursorF getCursor
+  , DefsUsesF getDefsUses
+  , FindNamesF findNames
+) {
```

Comment:
A more Rascal-y way would maybe be to define failing variants of these functions here, and extend them in the language specific module?